framework/base/Controller.php

```diff
@@ -137,33 +160,37 @@ public function runAction($id, $params = [])
         $oldAction = $this->action;
         $this->action = $action;
 
-        $modules = [];
-        $runAction = true;
-
-        // call beforeAction on modules
-        foreach ($this->getModules() as $module) {
-            if ($module->beforeAction($action)) {
-                array_unshift($modules, $module);
-            } else {
-                $runAction = false;
-                break;
+        $result = Yii::$app->getMiddlewareDispatcher()->dispatch($request, $this->middleware, function () use ($action, $params) {
```

Comment:
Could be moved out of controller to upper level (router which is Application in our case).